test/Performance/BuildPerformanceTest.cs

```diff
@@ -0,0 +1,249 @@
+                }
+            }
+        }
+
+        [Benchmark]
+        public void IncrementalRebuildWithRootChangedInGraph_SingleTargetGraph() =>
+            IncrementalRebuildWithRootChangedInGraph(CreateTestInstances(SingleTargetGraph));
+
+        [Benchmark]
+        public void IncrementalRebuildWithRootChangedInGraph_TwoTargetGraph() =>
+            IncrementalRebuildWithRootChangedInGraph(CreateTestInstances(TwoTargetGraph));
+
+        [Benchmark]
+        public void IncrementalRebuildWithRootChangedInGraph_TwoTargetGraphLarge() =>
+            IncrementalRebuildWithRootChangedInGraph(CreateTestInstances(TwoTargetGraphLarge));
+
+        public void IncrementalRebuildWithRootChangedInGraph(TestInstance[] instances)
+        {
+            var instance = instances[0];
+
+            Build(instance.TestRoot);
+
+            foreach (var iteration in Benchmark.Iterations)
+            {
+                using (iteration.StartMeasurement())
+                {
+                    Build(instance.TestRoot);
+                }
+                RemoveBin(instance.TestRoot);
+            }
+        }
+
+        [Benchmark]
+        public void IncrementalRebuildWithLastChangedInGraph_SingleTargetGraph() =>
+            IncrementalRebuildWithLastChangedInGraph(CreateTestInstances(SingleTargetGraph));
+
+        [Benchmark]
+        public void IncrementalRebuildWithLastChangedInGraph_TwoTargetGraph() =>
+            IncrementalRebuildWithLastChangedInGraph(CreateTestInstances(TwoTargetGraph));
+
+        [Benchmark]
+        public void IncrementalRebuildWithLastChangedInGraph_TwoTargetGraphLarge() =>
+            IncrementalRebuildWithLastChangedInGraph(CreateTestInstances(TwoTargetGraphLarge));
+
+        public void IncrementalRebuildWithLastChangedInGraph(TestInstance[] instances)
+        {
+            var instance = instances[0];
+
+            Build(instance.TestRoot);
+
+            foreach (var iteration in Benchmark.Iterations)
+            {
+                using (iteration.StartMeasurement())
+                {
+                    Build(instance.TestRoot);
+                }
+                RemoveBin(instances.Last().TestRoot);
+            }
+        }
+
+
+        [Benchmark]
+        public void IncrementalSkipAllNoDependenciesInGraph_SingleTargetGraph() =>
+           IncrementalSkipAllNoDependenciesInGraph(CreateTestInstances(SingleTargetGraph));
+
+        [Benchmark]
+        public void IncrementalSkipAllNoDependenciesInGraph_TwoTargetGraph() =>
+            IncrementalSkipAllNoDependenciesInGraph(CreateTestInstances(TwoTargetGraph));
+
+        [Benchmark]
+        public void IncrementalSkipAllNoDependenciesInGraph_TwoTargetGraphLarge() =>
+            IncrementalSkipAllNoDependenciesInGraph(CreateTestInstances(TwoTargetGraphLarge));
+
+        public void IncrementalSkipAllNoDependenciesInGraph(TestInstance[] instances)
+        {
+            var instance = instances[0];
+
+            Build(instance.TestRoot);
```

Comment:
Do you need to compile all of the projects before the first iteration?  Otherwise I'd expect all but the first to be compiled the first time around.